### PR TITLE
[LLM] Optimize ChatGLM2 kv_cache to support beam_search on ARC

### PR DESCRIPTION
**Pull request by sgwhat**
## Description
This PR supports ChatGLM2 model to generate long sequence with beam_search on ARC.

### 1. Why the change?

Previous implementation will meet oom trouble when prompt is too long, and cannot generate outputs.

### 2. User API changes

N/A

### 3. Summary of the change 
1. Optimize kv_cache size to reduce memory usage.
2. Keep `key_layer` and `value_layer` align to generate correct output.

### 4. How to test?
- [x] Unit test

See benchmark results in https://github.com/analytics-zoo/nano/issues/816.

### CODE REVIEW

File: python/llm/src/bigdl/llm/transformers/models/chatglm2.py [hunk 1]
```diff
@@ -246,13 +246,13 @@ def chatglm2_attention_forward_8eb45c(
         key_layer = key_layer.permute(1, 2, 0, 3).unsqueeze(-3)  # [bs, nh/k, sl, hn]
         key_layer = key_layer.expand(-1, -1, query_group_size, -1, -1)
         key_layer = key_layer.contiguous().view((batch_size,
```
Comment (yangw1234) [inline python/llm/src/bigdl/llm/transformers/models/chatglm2.py:248]: Looks like the `key_layer` here is of shape `[batch_size, num_multi_query_groups_per_partition, query_group_size, sequence_length, hidden_size_per_attention_head]`, how can it be viewed as `[batch_size, num_multi_query_groups_per_partition, sequene_length, hidden_size_per_attention_head]` if `query_group_size` is not 1.

File: python/llm/src/bigdl/llm/transformers/models/chatglm2.py [hunk 2]
```diff
@@ -293,13 +293,30 @@ def chatglm2_attention_forward_8eb45c(
...
+        cache_value_layer = value_layer
     else:
         kv_cache = None
 
     # ==================================
     # core attention computation
     # ==================================
+    if self.multi_query_attention:
+        key_layer = key_layer.unsqueeze(-3)
+        key_layer = key_layer.expand(-1, -1, 16, -1, -1)
```
[the remark this answers is not in the document]
Comment (jason-dai) [reply python/llm/src/bigdl/llm/transformers/models/chatglm2.py:304]: Need to add some comments

[the remark this answers is not in the document]
Comment (yangw1234) [reply python/llm/src/bigdl/llm/transformers/models/chatglm2.py]: > expend `key_layer` size from `[x, num_multi_query_groups_per_partition, x, x]` to `[x, num_attention_heads_per_partition, x, x]` for only core attention computation.

It is probably not a good idea to hard code the value here.

File: python/llm/src/bigdl/llm/transformers/models/chatglm2.py [hunk 3]
```diff
@@ -240,8 +240,25 @@ def chatglm2_attention_forward_8eb45c(
             key_layer = apply_rotary_pos_emb_chatglm(key_layer, rotary_pos_emb)
 
     if self.multi_query_attention:
-        key_layer = key_layer.permute(1, 2, 0, 3)  # [bs, nh/k, sl, hn]
-        value_layer = value_layer.permute(1, 2, 0, 3)  # [bs, nh/k, sl, hn]
+        if device.type == "xpu":
+            key_layer = key_layer.permute(1, 2, 0, 3)  # [bs, nh/k, sl, hn]
+            value_layer = value_layer.permute(1, 2, 0, 3)  # [bs, nh/k, sl, hn]
```
[the remark this answers is not in the document]
Comment (sgwhat) [reply python/llm/src/bigdl/llm/transformers/models/chatglm2.py:245]: Sure.

File: python/llm/src/bigdl/llm/transformers/models/chatglm2.py [hunk 4]
```diff
@@ -240,21 +240,25 @@ def chatglm2_attention_forward_8eb45c(
...
-        key_layer = key_layer.permute(1, 2, 0, 3).unsqueeze(-3)  # [bs, nh/k, sl, hn]
-        key_layer = key_layer.expand(-1, -1, query_group_size, -1, -1)
-        key_layer = key_layer.contiguous().view((batch_size,
-                                                 self.num_attention_heads_per_partition,
-                                                 key_length,
-                                                 self.hidden_size_per_attention_head))
-        value_layer = value_layer.permute(1, 2, 0, 3).unsqueeze(-3)
-        value_layer = value_layer.expand(-1, -1, query_group_size, -1, -1)
-        value_layer = value_layer.contiguous().view((batch_size,
+        if device.type == "xpu" and batch_size > 1:
```
Comment (yangw1234) [inline python/llm/src/bigdl/llm/transformers/models/chatglm2.py:243]: I think it is better to add some comments on why `device.type == "xpu" and batch_size > 1` is a special case that needs to be handled differently.

Comment (sgwhat) [reply python/llm/src/bigdl/llm/transformers/models/chatglm2.py:243]: Added.

File: python/llm/src/bigdl/llm/transformers/models/chatglm2.py [hunk 5]
```diff
@@ -265,13 +271,24 @@ def chatglm2_attention_forward_8eb45c(
...
-            new_cache_k, new_cache_v = extend_kv_cache(batch_size,
-                                                       self.num_attention_heads_per_partition,
-                                                       self.hidden_size_per_attention_head,
-                                                       past_length,
-                                                       max_cache_length,
-                                                       dtype=query_layer.dtype,
-                                                       device=device)
+            if device.type == "xpu" and batch_size > 1:
+                # Optimize memory usage and support beam_search on gpu.
+                # If you use beam_search for generation, the batch_size will be larger than 1.
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/models/chatglm2.py:276]: I think you need to explain how `batch_size > 1` is handled differently from `batch_size = 1`

**Review by hkvision — APPROVED** [review]

**Pull request merged**